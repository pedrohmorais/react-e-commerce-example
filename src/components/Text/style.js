import styled, { css } from 'styled-components';
import theme from 'styled-theming';
import { fontFamily } from 'theme';

const FONT_WEIGHT_REGULAR = 400;
const FONT_WEIGHT_SEMI = 600;
const FONT_WEIGHT_BOLD = 700;

const paragraphSize = theme.variants('mode', 'size', {
  tiny: { main: '10px' },
  small: { main: '12px' },
  medium: { main: '14px' },
  big: { main: '16px' },
});

const headingSize = theme.variants('mode', 'level', {
  1: { main: '25px' },
  2: { main: '22px' },
  3: { main: '16px' },
});

const fontWeight = theme.variants('mode', 'weight', {
  regular: { main: FONT_WEIGHT_REGULAR },
  semi: { main: FONT_WEIGHT_SEMI },
  bold: { main: FONT_WEIGHT_BOLD },
});

const textStyle = css`
  ${fontFamily}

  font-weight: ${fontWeight};
  font-style: ${({ italic }) => (italic ? 'italic' : 'normal')};
  text-transform: ${({ transform }) => transform};
`;

export const paragraphStyle = css`
  ${textStyle};
  font-size: ${paragraphSize};
`;

export const StyledParagraph = styled.p`
  ${paragraphStyle};
`;

export const StyledHeading = styled.h1`
  ${textStyle};
  font-size: ${headingSize};
`;

export const StyledInline = styled.span`
  ${paragraphStyle};
`;
