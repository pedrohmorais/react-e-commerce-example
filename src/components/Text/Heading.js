import React from 'react';
import {
  bool, node, oneOf, oneOfType, string,
} from 'prop-types';
import { StyledHeading } from 'components/Text/style';

const defaultProps = {
  level: '1',
  weight: 'semi',
  transform: 'none',
  italic: false,
  className: '',
};

const propTypes = {
  children: oneOfType([node, string]).isRequired,
  level: string,
  transform: oneOf(['none', 'capitalize', 'lowercase', 'uppercase']),
  weight: oneOf(['regular', 'semi', 'bold']),
  italic: bool,
  className: string,
};

const Heading = ({
  children, level, weight, transform, className, italic,
}) => {
  let HeadingLevel;

  switch (level) {
    case '2':
      HeadingLevel = StyledHeading.withComponent('h2');
      break;
    case '3':
      HeadingLevel = StyledHeading.withComponent('h3');
      break;
    default:
      HeadingLevel = StyledHeading;
      break;
  }

  return (
    <HeadingLevel
      level={level}
      weight={weight}
      transform={transform}
      className={className}
      italic={italic}
    >
      {children}
    </HeadingLevel>
  );
};

Heading.defaultProps = defaultProps;
Heading.propTypes = propTypes;

export default Heading;
