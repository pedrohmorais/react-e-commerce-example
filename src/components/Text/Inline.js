import React from 'react';
import {
  bool, node, oneOf, oneOfType, string,
} from 'prop-types';
import { StyledInline } from 'components/Text/style';

const defaultProps = {
  size: 'medium',
  weight: 'regular',
  transform: 'none',
  className: '',
  italic: false,
};

const propTypes = {
  children: oneOfType([node, string]).isRequired,
  size: oneOf(['tiny', 'small', 'medium', 'big']),
  weight: oneOf(['regular', 'semi', 'bold']),
  transform: oneOf(['none', 'capitalize', 'lowercase', 'uppercase']),
  className: string,
  italic: bool,
};

const Inline = ({
  children, size, weight, transform, className, italic,
}) => (
  <StyledInline
    size={size}
    weight={weight}
    transform={transform}
    className={className}
    italic={italic}
  >
    {children}
  </StyledInline>
);

Inline.defaultProps = defaultProps;
Inline.propTypes = propTypes;

export default Inline;
