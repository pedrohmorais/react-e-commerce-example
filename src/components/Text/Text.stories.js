import React from 'react';
import { storiesOf } from '@storybook/react';
import { select, text, boolean } from '@storybook/addon-knobs';

import Text from 'components/Text';

const children = () => text('children', 'Hello storybook');
const size = () => select('size', ['tiny', 'small', 'medium'], 'medium');
const transform = () => select('transform', ['none', 'capitalize', 'lowercase', 'uppercase'], 'none');
const weight = defaultValue => select('weight', ['regular', 'semi', 'bold'], defaultValue);
const italic = () => boolean('italic', false);

storiesOf('Text', module)
  .add('index', () => (
    <Text weight={weight('regular')} size={size()} transform={transform()} italic={italic()}>
      {children()}
    </Text>
  ))
  .add('Inline', () => (
    <Text.Inline weight={weight('regular')} size={size()} transform={transform()} italic={italic()}>
      {children()}
    </Text.Inline>
  ))
  .add('Heading', () => (
    <Text.Heading
      weight={weight('semi')}
      level={select('level', ['1', '2', '3'], '1')}
      transform={transform()}
      italic={italic()}
    >
      {children()}
    </Text.Heading>
  ));
