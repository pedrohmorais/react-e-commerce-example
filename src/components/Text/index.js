import React from 'react';
import {
  bool, node, oneOf, oneOfType, string,
} from 'prop-types';

import Heading from 'components/Text/Heading';
import Inline from 'components/Text/Inline';
import { StyledParagraph } from 'components/Text/style';

const defaultProps = {
  size: 'medium',
  weight: 'regular',
  transform: 'none',
  italic: false,
  className: '',
};

const propTypes = {
  children: oneOfType([node, string]).isRequired,
  size: oneOf(['tiny', 'small', 'medium', 'big']),
  weight: oneOf(['regular', 'semi', 'bold']),
  transform: oneOf(['none', 'capitalize', 'lowercase', 'uppercase']),
  italic: bool,
  className: string,
};

const Text = ({
  children, size, weight, transform, italic, className,
}) => (
  <StyledParagraph
    size={size}
    weight={weight}
    transform={transform}
    italic={italic}
    className={className}
  >
    {children}
  </StyledParagraph>
);

Text.defaultProps = defaultProps;
Text.propTypes = propTypes;
Text.Heading = Heading;
Text.Inline = Inline;

export default Text;
