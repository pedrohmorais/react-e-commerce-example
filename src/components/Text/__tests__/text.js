import React from 'react';
import { shallow } from 'enzyme';

import Text from 'components/Text';

describe('<Text />', () => {
  describe('Paragraph', () => {
    it('should match snapshot', () => {
      const Paragraph = (
        <Text>
Hello World!
        </Text>
      );

      expect(shallow(Paragraph)).toMatchSnapshot();
    });
  });

  describe('Inline', () => {
    it('should match snapshot', () => {
      const Inline = (
        <Text.Inline>
Hello World!
        </Text.Inline>
      );

      expect(shallow(Inline)).toMatchSnapshot();
    });
  });

  describe('Heading', () => {
    it('H1 should match snapshot', () => {
      const H1 = (
        <Text.Heading>
Hello World!
        </Text.Heading>
      );

      expect(shallow(H1)).toMatchSnapshot();
    });

    it('H2 should match snapshot', () => {
      const H2 = (
        <Text.Heading level="2">
Hello World!
        </Text.Heading>
      );

      expect(shallow(H2)).toMatchSnapshot();
    });

    it('H3 should match snapshot', () => {
      const H3 = (
        <Text.Heading level="3">
Hello World!
        </Text.Heading>
      );

      expect(shallow(H3)).toMatchSnapshot();
    });
  });
});
