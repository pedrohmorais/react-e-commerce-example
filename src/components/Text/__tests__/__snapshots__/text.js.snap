// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> Heading H1 should match snapshot 1`] = `
<StyledHeading
  className=""
  italic={false}
  level="1"
  transform="none"
  weight="semi"
>
  Hello World!
</StyledHeading>
`;

exports[`<Text /> Heading H2 should match snapshot 1`] = `
<StyledHeading
  className=""
  italic={false}
  level="2"
  transform="none"
  weight="semi"
>
  Hello World!
</StyledHeading>
`;

exports[`<Text /> Heading H3 should match snapshot 1`] = `
<StyledHeading
  className=""
  italic={false}
  level="3"
  transform="none"
  weight="semi"
>
  Hello World!
</StyledHeading>
`;

exports[`<Text /> Inline should match snapshot 1`] = `
<StyledInline
  className=""
  italic={false}
  size="medium"
  transform="none"
  weight="regular"
>
  Hello World!
</StyledInline>
`;

exports[`<Text /> Paragraph should match snapshot 1`] = `
<StyledParagraph
  className=""
  italic={false}
  size="medium"
  transform="none"
  weight="regular"
>
  Hello World!
</StyledParagraph>
`;
